### ISSUE TITLE
Downward API for resource limits

### ISSUE THREAD

**Issue by justinsb** [kubernetes/kubernetes#9473]
Is it possible inside a container to discover your resource limits, in particular the memory limit.

Apparently this is sometime put under `/sys/fs/cgroup/memory/memory.stat`, but that isn't present.  And /proc/meminfo is the memory of the host (AFAICT)

I would like, for example, to specify memory limits for things like memcache, Java, Node, Postgres, Mysql based on the resource limits given to me.  I could query the k8s API, but that seems very heavy.

**Comment by dchen1107** [comment]
This is discussed in https://github.com/GoogleCloudPlatform/kubernetes/issues/9356#issuecomment-110155706

**Comment by justinsb** [comment]
@dchen1107 - thanks for the link, but I don't think that applies here?  (Unless there's a trick I'm missing, which is not unlikely).  I know I can get the information from the API (via kubectl or curl), I was just hoping for a more direct way.  In my host I have /sys/fs/cgroup/memory/, and apparently this is usually available in Docker containers also, although I don't see it in my Docker k8s containers (on AWS or GCE).

**Comment by bgrant0607** [comment]
We call this "downward API". See also #386. cc @pmorie

**Comment by pmorie** [comment]
@bgrant0607 My bad, I somehow missed being tagged into this.

@justinsb I definitely think this should be available via the downward API.

**Comment by pmorie** [comment]
@justinsb I'm going to take a stab at this today.

**Comment by pmorie** [comment]
@justinsb Let me make sure we're on the same page about what you want before I write code.  It sounds to me like what you're asking for is: if you set the `ResourceRequirements` field of a container, you want to be able to consume that information via the downward API.  Accurate?

**Comment by justinsb** [comment]
@pmorie I think this would be useful for the downwards API.  But: my workaround is to simply list _all_ the pods and match podIP, and get my full pod info that way.  This means that the manifest doesn't require the user to include any additional information (i.e. the downward API mapping).  It's not pretty, and it isn't efficient because we can't currently filter by podIP, but I find it pretty tolerable.

**Comment by pmorie** [comment]
It seems like the biggest challenge to doing this is getting the API right.  Currently, the downward API only supports references to `Pod`-scoped fields, like so:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: dapi-test-pod
spec:
  containers:
    - name: test-container
      image: gcr.io/google_containers/busybox
      command: [ "/bin/sh", "-c", "env" ]
      env:
        - name: POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
  restartPolicy: Never
```

With this change we need to expand the scope of the API to container-scoped fields.  There are a couple different facets to consider:

1.  If you're consuming the API in environment variables, it might be natural to have a container-scoped field of the `EnvVarSource`
2.  If you're consuming the API in a volume (see https://github.com/GoogleCloudPlatform/kubernetes/pull/5093, which I hope will be merged soon), the scope in which you specify fields is pod scoped, so if you want to be able to consume resource limits in a volume, we need a way to address a container's resource limits in a `ObjectFieldSelector`

I'm not sure that this value makes a lot of sense to consume in an environment variable.

Any thoughts @smarterclayton @bgrant0607 ?

**Comment by pmorie** [comment]
Example of possible API:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: dapi-test-pod
spec:
  containers:
    - name: test-container
      image: gcr.io/google_containers/busybox
      command: [ "/bin/sh", "-c", "env" ]
      env:
        - name: CONTAINER_MEMORY
          valueFrom:
            containerFieldRef:
              fieldPath: "resources.limits.memory"
  restartPolicy: Never

**Comment by pmorie** [comment]
Some other little nits about this:

1.  The `Quantity` API type supports formatting, which is something you would conceivably want to apply to the quantities for this API
2.  The `ResourceList` type is actually a `map[ResourceType]resource.Quantity` -- what's right expression syntax we want for this?  Do we want to parse `resources.limits[memory]` in an api label conversion function?

**Comment by bgrant0607** [comment]
Perhaps @rjnagal, @vmarmol, or @vishh have thoughts about this. I'd prefer a non-Kubernetes-specific way to communicate resource info, if feasible. (I'd consider a kernel change to be infeasible, as an example, at least in the short term.)

**Comment by pmorie** [comment]
@bgrant0607 Can you elaborate on what you meant by non-Kubernetes-specific way?  I'm assuming you mean to decouple the container from the kubernetes API, but I want to verify.

**Comment by vmarmol** [comment]
Most of the resource isolation parameters are available through cgroups. Although some of these may require some translation to be useful (like shares). Even then it may depend on how the application sets up the containers. It may be good enough for now, but there are many details that won't work in a more "general" way.

**Comment by bgrant0607** [comment]
Yes, decoupled from the Kubernetes API, both from the precise schema and also from inherently Kubernetes-specific idioms. (Arguably pods don't have to be Kubernetes-specific, as a few other systems have adopted them.)

**Comment by thockin** [comment]
I don't know that I buy the idea that people will take "run this external
command" as a way to get container limits and other info.  It just doesn't
sound practical to me.  I also don't buy that going to the kernel is the
best way to do it.  I think we either offer a runtime API, a pre-pop file,
or we do something like what Paul is suggesting here.

This exposes some of what I was afraid of with fieldpath - it sounds really
general purpose, but without an anchor like "this container" it's not
really useful.

**Comment by bgrant0607** [comment]
Who suggested that we run an external command? Was that in another issue or PR?

**Comment by bgrant0607** [comment]
For the env var source, we could use a JSONPath-like expression -- see #11738. That might look something like: `.spec.containers[?(@.name=="test-container")].resources.limits.memory`. Maps are indistinguishable from struct fields in JSON. containerFieldRef would be more convenient, but would then no longer be a subset of ObjectReference. We could push the syntax into the field specification instead, like `$container.` or `...` or something.

**Comment by thockin** [comment]
$container is what I meant - providing contextual anchors for field paths
seems like the nicest API we could provide, though sort of bespoke.  The
default anchor would be implied by a leading dot or an explicit $ or $pod.
 $container would be the current container., etc...

**Comment by vishh** [comment]
Do we ever plan to support dynamic updates of compute resources?

**Comment by smarterclayton** [comment]
Yes, another issue #8157 exists for it

On Wed, Jul 29, 2015 at 1:43 PM, Vish Kannan <notifications@github.com>
wrote:

> Do we ever plan to support dynamic updates of compute resources?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/issues/9473#issuecomment-126032074>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by pmorie** [comment]
@bgrant0607 @thockin I like the contextual anchor idea.  I think something like that is needed in order to have a single expression field that is generally useful.  Otherwise we will have to differentiate fields in order to have different targets for different expressions.

**Comment by pmorie** [comment]
@vishh Good question -- perhaps this should be exposed solely via the volume plugin, which will support eventual consistency.

**Comment by vishh** [comment]
@pmorie: Have we considered exposing a versioned, read-only file, inside the container that exposes resource limits?

**Comment by pmorie** [comment]
@vishh Can you define 'versioned' in this context?

**Comment by vishh** [comment]
'versioned' to let us evolve the file format in the future and provide backwards compat.

**Comment by pmorie** [comment]
@vishh Got it.  I don't think anyone's talked about a versioned format, but I think versioning is a good idea.  #5093 will add a volume plugin for the fields usable with the existing downward API.  It seems like the versioned file format would be a special case for that plugin -- agree?

**Comment by vishh** [comment]
@pmorie: Yup!

**Comment by bgrant0607** [comment]
Our general principle with the downward API is that we should give the user enough control over the structure of the data that we don't need to version the format explicitly, separately from the cluster API version.

**Comment by lavalamp** [comment]
Considering that we may want to change this value at run time without restarting the container(s), it seems like the current downward API won't do the trick, since that's fixed at start-up.

**Comment by pmorie** [comment]
@lavalamp when #5093 merges we can expose this via the volume plugin.
Agree that values that may change at runtime are not suitable for being
exposed as envs.

**Comment by bgrant0607** [comment]
I didn't mean to imply that we should use env vars, just that we give the user control over the format and the information provided.

**Comment by derekwaynecarr** [comment]
This came to my attention from another Red Hat employee via mail, but seemed pertinent to this issue discussion since it provides a mechanism for containers to inspect its own limits.

>>> Just as a FYI --
> docker 1.8 mounts a container's own cgroups read-only inside the container
> by default, so the container process can inspect its own limits.
>
>
> [root@arch ~]# docker run -it --rm busybox sh
> / #
> / # ls -l /sys/fs/cgroup/
> total 0
> drwxr-xr-x    2 root     root             0 Aug 27 22:14 blkio
> drwxr-xr-x    2 root     root             0 Aug 27 22:14 cpu,cpuacct
> drwxr-xr-x    2 root     root             0 Aug 27 22:14 cpuset
> drwxr-xr-x    2 root     root             0 Aug 27 22:14 devices
> drwxr-xr-x    2 root     root             0 Aug 27 22:14 freezer
> drwxr-xr-x    2 root     root             0 Aug 27 22:14 memory
> drwxr-xr-x    2 root     root             0 Aug 27 22:14 net_cls
> drwxr-xr-x    2 root     root             0 Aug 27 22:14 systemd
> / #
> / # cat /sys/fs/cgroup/memory/memory.limit_in_bytes
> 9223372036854771712
> / # exit

**Comment by vishh** [comment]
Cgroups based interface will work for memory, but it won't work for CPU.

**Comment by pmorie** [comment]
@derekwaynecarr i wish it was part of a spec that spanned multiple
container runtimes.

**Comment by derekwaynecarr** [comment]
@vishh @pmorie - making it part of spec is useful, I just wanted to make sure that people knew there was an alternate method to get memory limits in the interim.

**Comment by thockin** [comment]
Let's also be clear - upstream kernel folks have, in the past, declared
this as dangerous and broken, and not guaranteed to be supported.

**Comment by derekwaynecarr** [comment]
@thockin - understood, just wanted to make sure that @justinsb knew that the original location he had been looking for the memory information should be populated in docker 1.8.  This is just one potential stop-gap until real kernel solution is available.

**Comment by vishh** [comment]
@pmorie: How about exposing the API server APIs via the kubelet in read-only mode?

**Comment by pmorie** [comment]
@vishh Do you need a kubeconfig and client in the container to use it?

**Comment by vishh** [comment]
I think so. If we ignore access control for now, since the APIs are
read-only, can we skip auth?

**Comment by vishh** [comment]
@pmorie: Were you able to make any progress?

**Comment by lavalamp** [comment]
This is a P1 issue-- what is the status? If no one is working on it, I will demote to P2.

**Comment by pmorie** [comment]
I haven't had time to work on this further, we should bump down to p2

**Comment by bgrant0607** [comment]
cc @dchen1107 @davidopp

**Comment by bgrant0607** [comment]
In case someone didn't notice it, the proposal is #24051

**Comment by ncdc** [comment]
Implemented in #24179

**Comment by pmorie** [comment]
Closed by #24179

**Comment by dragon9783** [comment]
```
.spec.containers[?(@.name=="test-container")].resources.limits.memory
```
LGTM
